### PR TITLE
Add conditional rendering of Login and Logout buttons in the UI.

### PR DESCRIPTION
**Pull request by gootla-manasa**
Depending on the availability of "access_token" in the state of "App" component, Login and Logout buttons in the header are toggled.

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -10,36 +10,36 @@ class App extends Component {
...
+  isLoggedIn = function () {
+    if (this.state && this.state.authenticationInfo && this.state.authenticationInfo.access_token) {
+      return true;
+    }
+    return false;
+  }
+
+  handleAuthenticationInfoUpdate = function (authResponse) {
+    if (!authResponse) {
+      this.setState({authenticationInfo: undefined})
```
Comment (s-maheshbabu) [inline src/App.js]: Add a log line here too. This is also an exceptional case.

File: src/App.js [hunk 2]
```diff
@@ -10,36 +10,36 @@ class App extends Component {
...
   }
 
-  handleAuthenticationInfoUpdate = function(authResponse) {
-    if (!authResponse || authResponse.error) {
-      console.log(
-        "Encountered an error on login: " +
-          util.inspect(authResponse, { showHidden: true, depth: null })
-      );
+  isLoggedIn = function () {
+    if (this.state && this.state.authenticationInfo && this.state.authenticationInfo.access_token) {
```
Comment (s-maheshbabu) [inline src/App.js]: Use hasIn from immutable.js to make this more readable.

Comment (s-maheshbabu) [reply src/App.js]: More over can't this be written as return (X && Y && Z); ?

File: src/App.js [hunk 3]
```diff
@@ -10,36 +10,36 @@ class App extends Component {
...
+    } else if (authResponse.error) {
+      console.log("Encountered an error on login: " + util.inspect(authResponse, {showHidden: true, depth: null}));
     } else {
-      this.setState({
-        authenticationInfo: {
-          access_token: authResponse.access_token,
-          expires_in: authResponse.expires_in
-        }
-      });
+      console.log(authResponse);
```
Comment (s-maheshbabu) [inline src/App.js]: Didn't we decide not to log successful auth responses in a previous CR?

File: src/App.test.js [hunk 4]
```diff
@@ -16,9 +16,17 @@ it("renders without crashing", () => {
...
-it("snapshot testing that it renders correctly", () => {
+it("renders correctly (snapshot testing)", () => {
   const wrapper = mount(<App />);
   expect(wrapper).toMatchSnapshot();
+  wrapper.unmount();
+});
+
+it("renders LogoutButton when authenticationInfo state changes", () => {
+  const wrapper = mount(<App />);
+  wrapper.setState({authenticationInfo: {access_token: "some_access_token", expires_in: "30"}});
```
Comment (s-maheshbabu) [inline src/App.test.js]: I don't think this spec is correct - "renders LogoutButton when authenticationInfo state changes". We don't want to show Logout if the authInfo changed from a valid object to undefined, for example. 

It should say something like "renders LogoutButton when state changes to contain a valid authInfo."

Then, I think we should make the test more explicit. How about assign state to undefined first and then set a valid auth info?

File: src/App.test.js [hunk 5]
```diff
@@ -16,9 +16,17 @@ it("renders without crashing", () => {
   shallow(<App />);
 });
```
Comment (s-maheshbabu) [inline src/App.test.js]: Why is there no spec that authInfo is valid only if expires_time is within range?

File: src/App.test.js (hunk 4 above)
Comment (s-maheshbabu) [reply src/App.test.js]: On second thoughts, aren't you testing beyond the boundaries of your 'unit'? Shouldn't you just test that isLogged value is propagated to a 'shallow' component? You can always change the state to contain different authInfo objects to check that the right isLogged value is generated.

File: src/Header.test.js [hunk 6]
```diff
@@ -2,20 +2,36 @@ import React from "react";
...
 it("renders correctly (snapshot testing)", () => {
-  const wrapper = mount(<Header />);
+  mockIsLoggedIn.mockReturnValueOnce(false);
+  const wrapper = mount(<Header isLoggedIn={mockIsLoggedIn} />);
+  expect(wrapper).toMatchSnapshot();
+
+  wrapper.unmount();
+});
+
+it("renders correctly logout (snapshot testing)", () => {
```
Comment (s-maheshbabu) [inline src/Header.test.js]: should you rename the previous test to 'login'.

Comment (s-maheshbabu) [reply src/Header.test.js]: Moreover, I don't think this is being tested right. Don't both the above tests generate the same snapshot file with something like <Header isLoggedIn=[Function]/>?

I don't know how to test this but this doesn't look right. Check the snapshot files.

File: src/Header.test.js [hunk 7]
```diff
@@ -2,20 +2,36 @@ import React from "react";
...
+  const wrapper = mount(<Header isLoggedIn={() => mockIsLoggedIn} />);
   expect(wrapper).toMatchSnapshot();
 
   wrapper.unmount();
 });
 
 it("verifies that updateAuthenticationInfo function is passed to LoginControl component", () => {
   const mockCallback = jest.fn();
-  const header = shallow(<Header updateAuthenticationInfo={mockCallback} />);
+  const header = shallow(<Header isLoggedIn={jest.fn()} updateAuthenticationInfo={mockCallback} />);
```
Comment (s-maheshbabu) [inline src/Header.test.js]: Well, this is probably how it should be tested. Just like updateAuthenticationInfo prop is being tested.

File: src/LoginControl.test.js [hunk 8]
```diff
@@ -23,28 +17,50 @@ beforeEach(() => {
...
-  const wrapper = mount(<LoginControl />);
+it("renders LoginButton when isLoggedIn returns false (snapshot testing)", () => {
+  mockIsLoggedIn.mockReturnValueOnce(false);
+  const wrapper = mount(<LoginControl isLoggedIn={mockIsLoggedIn} />);
+  expect(wrapper).toMatchSnapshot();
+
+  wrapper.unmount();
+});
+
+it("renders LogoutButton when isLoggedIn returns true (snapshot testing)", () => {
```
Comment (s-maheshbabu) [inline src/LoginControl.test.js]: previous test to be renamed because now it tests a specific case?

File: src/LoginControl.test.js [hunk 9]
```diff
@@ -23,28 +17,50 @@ beforeEach(() => {
...
+  mockIsLoggedIn.mockReturnValueOnce(true);
+  const wrapper = mount(<LoginControl isLoggedIn={mockIsLoggedIn} />);
   expect(wrapper).toMatchSnapshot();
 
   wrapper.unmount();
 });
```
Comment (s-maheshbabu) [inline src/LoginControl.test.js]: spec doesn't contain the cases when isLoggedIn is not present in the props or when it is not a function.

File: src/LogoutButton.test.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import React from "react";
+import {shallow, mount} from "enzyme";
+import LogoutButton from "./LogoutButton";
+
+it("renders LogoutButton without crashing", () => {
+  mount(<LogoutButton />);
+});
+
+it("renders correctly (snapshot testing)", () => {
```
Comment (s-maheshbabu) [inline src/LogoutButton.test.js]: When we have snapshot testing, is the 'renders without crashing' needed? Please read about it to see what the best practice is.

File: src/LogoutButton.test.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  const wrapper = mount(<LogoutButton />);
+  expect(wrapper).toMatchSnapshot();
+
+  wrapper.unmount();
+});
+
+it("verifies that the callback function is called when login button is clicked", () => {
```
Comment (s-maheshbabu) [inline src/LogoutButton.test.js]: logout button?

**Comment by gootla-manasa** [comment]
Dismissing this request in the wake of the requested changes.

**Pull request closed without merging**